### PR TITLE
Add strong types for JSON.stringify

### PR DESCRIPTION
**Pull request by sstur**
The built-in types for `JSON.stringify()` are not correct/sound because `JSON.stringify()` will return `undefined` if `undefined` is passed in. Many devs don't know this and TS types don't reflect this possibility, resulting in a false sense of null safety and often runtime exceptions in prod.

This fixes that using conditional types. If there's a better way to express this other than conditional types please let me know and I'll update this PR to reflect it.

Thanks!

### CODE REVIEW

File: src/tests/json-stringify.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { doNotExecute, Equal, Expect } from "./utils";
+
+doNotExecute(() => {
+  const result = JSON.stringify(undefined);
+  type tests = [Expect<Equal<typeof result, string | undefined>>];
```
Comment (mattpocock) [inline src/tests/json-stringify.ts:5]: `result` here should probably be `undefined`!

File: src/tests/json-stringify.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+});
+
+doNotExecute(() => {
+  const result = JSON.stringify(5 as undefined | number);
+  type tests = [Expect<Equal<typeof result, string | undefined>>];
+});
+
+doNotExecute(() => {
+  const result = JSON.stringify(undefined as any);
```
Comment (mattpocock) [inline src/tests/json-stringify.ts:14]: Should result be `string` here?

File: src/tests/json-stringify.ts (hunk 1 above)
Comment (sstur) [reply src/tests/json-stringify.ts:5]: You're right. I hesitated to use a nested conditional type here for this edge case, opting for the simpler typedef (that is still sound), but technically yes we can tighten this up to be undefined if we are 100% sure the arg passed in is undefined (and not just `extends undefined` since any would fit that bill).

File: src/tests/json-stringify.ts (hunk 2 above)
Comment (sstur) [reply src/tests/json-stringify.ts:14]: No, I don't think so. If `any` is received then we can't assume anything about the param, so we should return the broadest possible type `string | undefined` but correct me if I'm wrong.

File: src/tests/json-stringify.ts (hunk 1 above)
Comment (mattpocock) [reply src/tests/json-stringify.ts:5]: I think an extra overload makes sense here.

File: src/tests/json-stringify.ts (hunk 2 above)
Comment (mattpocock) [reply src/tests/json-stringify.ts:14]: Agree with your logic here. Could you add a descriptive comment to that effect?

File: src/entrypoints/json-stringify.d.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+interface JSON {
+  /**
+   * Converts a JavaScript value to a JavaScript Object Notation (JSON) string.
+   * @param value A JavaScript value, usually an object or array, to be converted.
+   * @param replacer A function that transforms the results.
+   * @param space Adds indentation, white space, and line break characters to the return-value JSON text to make it easier to read.
+   */
+  stringify<T = any>(
```
Comment (sstur) [inline src/entrypoints/json-stringify.d.ts:8]: BTW, is it useless to have the `= any` here? I'm just realizing I probably can remove this.

File: src/tests/json-stringify.ts (hunk 1 above)
Comment (sstur) [reply src/tests/json-stringify.ts]: Done.

**Comment by sstur** [comment]
Let me know if anything else I should add to this PR (update the README? include in recommended preset?). Thanks!

File: src/tests/json-stringify.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  type tests = [Expect<Equal<typeof result, string | undefined>>];
+});
+
+doNotExecute(() => {
+  const result = JSON.stringify(undefined as unknown);
+  type tests = [Expect<Equal<typeof result, string | undefined>>];
+});
+
+doNotExecute(() => {
+  const value: null | string | number | boolean | object | Array<any> = null;
```
Comment (zaverden) [inline src/tests/json-stringify.ts:26]: despite you put a type for a value variable, TS narrows it to `null` ([see here](https://tsplay.dev/mpjMpw)). update it with `as` to get expected type

```suggestion
  const value = null as (null | string | number | boolean | object | Array<any>);
```

File: src/tests/json-stringify.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import { doNotExecute, Equal, Expect } from "./utils";
+
+doNotExecute(() => {
+  const result = JSON.stringify(undefined);
+  type tests = [Expect<Equal<typeof result, undefined>>];
+});
```
Comment (zaverden) [inline src/tests/json-stringify.ts:6]: functions are also resolved as undefined

```typescript
doNotExecute(() => {
  const fn = () => {};
  const result = JSON.stringify(fn);
  type tests = [Expect<Equal<typeof result, undefined>>];
});
```

Comment (sstur) [reply src/tests/json-stringify.ts:6]: Oh, this is a good catch! I didn't even realize that. Checking into it...

File: src/tests/json-stringify.ts (hunk 4 above)
Comment (sstur) [reply src/tests/json-stringify.ts]: Done.

File: src/tests/json-stringify.ts (hunk 5 above)
Comment (sstur) [reply src/tests/json-stringify.ts:6]: Hmm, so it turns out this strange behavior of JSON.stringify() treating a function as undefined sorta throws a wrench in the works, because a function can be assigned to `Record<string, any>`, meaning it's hard to be certain that we have an object that is not a function.

I've update this PR to account for that possibility but be aware this makes it so that many more situations will fall back to `string | undefined` vs just `string`.

I think we should consider ignoring the "function treated as undefined" edge case and give up some soundness for developer ergonomics. It would still be far more correct than the in-built types, but would not be perfectly sound (choosing pragmatism over soundness is the long standing posture of TS anyway).

Thoughts @mattpocock?

Comment (sstur) [reply src/tests/json-stringify.ts:6]: To put this another way:
Many legit values (that are never intended to be a function) in the real world will likely have the type `Record<string, any>` and if we got for soundest possible types for `JSON.stringify()` then passing in those real-world values will result in `string | undefined` being returned, likely confusing the average TS dev who doesn't realize why. Is this a tradeoff we want to make?

Comment (mattpocock) [reply src/tests/json-stringify.ts:6]: Could you add an overload to catch it being a function? And then only return undefined.

**Comment by KotlinIsland** [comment]
- Typescripts issue for this is here https://github.com/microsoft/TypeScript/issues/18879

**Comment by sstur** [comment]
Thanks @KotlinIsland! As you probably discovered, a correct typedef for JSON.stringify was attempted and even [landed](https://github.com/microsoft/TypeScript/pull/51897) in TS core, but it was [rolled back](https://github.com/microsoft/TypeScript/pull/52307) because it broke too much real-world code. According to RyanCavanaugh in his comment [here](https://github.com/microsoft/TypeScript/issues/18879#issuecomment-1399758565) this is not likely to make it into core.

I suppose stuff like this is a big part of why ts-reset exists!

File: src/tests/json-stringify.ts (hunk 5 above)
Comment (aaditmshah) [reply src/tests/json-stringify.ts:6]: I think that the return type should always be `string | undefined`. If the developer is confident that the result is not `undefined` then they can always use the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator).

```typescript
const maybeString = JSON.stringify(foo); // string | undefined

const definitelyString = JSON.stringify(bar)!; // string
```

IMHO, soundness is more important than pragmatism. There shouldn't be any guesswork involved. Always choose the safe option over the convenient option, and allow developers to explicitly bypass the safety if they are confident.

Comment (aaditmshah) [reply src/tests/json-stringify.ts:6]: By the way, symbols are also resolved as `undefined`.

Comment (KotlinIsland) [reply src/tests/json-stringify.ts:6]: > I think that the return type should always be `string | undefined`.

> IMHO, soundness is more important than pragmatism.

I agree, but there are known cases where it will only return `string`, so those cases should be accounted for.

Comment (aaditmshah) [reply src/tests/json-stringify.ts:6]: > I agree, but there are known cases where it will only return `string`, so those cases should be accounted for.

There are two problems with specializing the return type to only `string` or only `undefined`.

1. It's not foolproof because TypeScript is unsound. Hence, the input type at compile time might not match the input type at runtime. For example, consider the following code.

   ```typescript
   // This is a contrived example. However, because TypeScript is unsound this is a possibility.
   const stringifyContentType = (headers: Record<string, string>) => {
     // Suppose a novice developer uses an unchecked index access.
     // The type of `contentType` is `string` because `noUncheckedIndexedAccess` is not enabled.
     // However, at runtime the value of `contentType` might be `undefined`.
     const contentType = headers["Content-Type"];

     // Since the type of `contentType` is `string`, we specialize the result type to `string`.
     // However, this is incorrect because the type of `contentType` is incorrect.
     return JSON.stringify(contentType);
   };
   ```

   Now, you might argue that if the input type is incorrect then it's all right for the output type to be incorrect too. After all, garbage in, garbage out. Fair enough, but where do you draw the line?

   If it's all right to assume that a value typed `string` is not an `undefined` at runtime, then is it not also all right to assume that a value typed `Record<string, any>` is not a function at runtime?

2. Is it worth the effort? Specializing the return type to only `string` or only `undefined` requires complex type definitions. Personally, I don't mind complex type definitions if it can guarantee reasonable type-safety. However, as we saw earlier if you can't trust the input type then you also can't trust the specialized return type.

   Although I prize type-safety above everything else, other people might be more pragmatic. So, do these complex type definitions, which specialize the return type to only `string` or only `undefined`, benefit people who are pragmatic? Let's consider the specialized return types on a case-by-case basis. We'll trust that the return types are correct.

   1. If the return type is specialized to `undefined`, then that's not useful at all. The developer might as well delete any calls to `JSON.stringify` that return only `undefined`.
   2. If the return type is specialized to `string`, then the developer is lulled into a false sense of security. Sure, the result might actually be a `string`, but what if it isn't? As I described earlier, it is indeed possible for the result to be `undefined` because TypeScript is unsound.

So, it seems to me that the best option is to never specialize the return type. Let the return type be `string | undefined`. If the user is confident that the return type is indeed a `string` then they can explicitly tell the type checker so by using the non-null assertion operator.

In summary, I believe that never specializing the return type is the best option because:

1. It's always safe.
2. Developers can always use the non-null assertion operator to bypass this safety.
3. No need for complex type definitions. Keep it straight and simple.

Comment (aaditmshah) [reply src/tests/json-stringify.ts:6]: By the way, here are three addition examples of using `JSON.stringify` where the return type is incorrectly specialized to `string`. All three values, `foo`, `bar`, and `baz` are `undefined` because of the special `toJSON` method.

```typescript
const foo = JSON.stringify({ toJSON: () => {} });
const bar = JSON.stringify({ toJSON: () => () => {} });
const baz = JSON.stringify({ toJSON: () => Symbol() });
```

As you can see, it's easier to make the return type always be `string | undefined` than it is to create specialized type definitions for all the possible edge cases.

Comment (KotlinIsland) [reply src/tests/json-stringify.ts:6]: > is it not also all right to assume that a value typed `Record<string, any>` is not a `function` at runtime?

It is right to assume that a value typed `Record<string, any>` could be a function at runtime:
```ts
const a: Record<string, any> = () => {}
```
[playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAhgLhgJQKagE4BMA80MCWYA5gDTxgCeAfDALwwAUAlPbQN4C+QA)

> If the return type is specialized to `undefined`, then that's not useful at all. The developer might as well delete any calls to `JSON.stringify` that return only `undefined`.

This would actually be useful, it will alert the developer to a case when they have accidentally used `stringify` in an invalid case.


> If the return type is specialized to`string`, then the developer is lulled into a false sense of security. Sure, the result might actually be a `string`, but what if it isn't? As I described earlier, it is indeed possible for the result to be `undefined` because TypeScript is unsound.

That argument could be applied to literally any value:

```ts
function foo(x: string) {
    // we say x is a string, but someone passed null to this function!
    console.log(x.toLowerCase());
}
const a: string | null = null
foo(a!)
```

> As you can see, it's easier to make the return type always be `string | undefined` than it is to create specialized type definitions for all the possible edge cases.

I don't think it's a far-fetched idea to specialize this function, even the `toJSON` case could be accounted for in the conditional return type.

Comment (aaditmshah) [reply src/tests/json-stringify.ts:6]: > > is it not also all right to assume that a value typed `Record<string, any>` is not a `function` at runtime?
>
> It is right to assume that a value typed `Record<string, any>` could be a function at runtime:
> 
> ```ts
> const a: Record<string, any> = () => {}
> ```

There are two problems with this argument.

1. The type `Record<string, any>` uses `any`, which disables type checking. Change the type to `Record<string, unknown>` instead and you'll see the type error.

   For any type `A`, the type `Record<string, A>` should be assignable to `Record<string, unknown>` because the type `A` is in a covariant position. The type `() => void` is not assignable to `Record<string, unknown>`. This means that `() => void` is not assignable to `Record<string, A>` for some type `A`, i.e. _functions are not records_.

2. `() => void` being assignable to `Record<string, any>` should be considered a bug. In fact, I went ahead and raised a bug issue for it.

   https://github.com/microsoft/TypeScript/issues/53484

   Hopefully, we'll get some clarity on why this issue is occurring. If we get an official answer that this is indeed the expected behavior then I'll be happy to concede this point to you.

That being said, I can see how `Object.assign(() => {}, someRecord)` would have the type `(() => void) & Record<string, unknown>` which is indeed assignable to `Record<string, unknown>`. Thus, `Record<string, unknown>` could indeed be a function at runtime. All the more reason to not specialize the return type of `JSON.stringify`.

> > If the return type is specialized to `undefined`, then that's not useful at all. The developer might as well delete any calls to `JSON.stringify` that return only `undefined`.
> 
> This would actually be useful, it will alert the developer to a case when they have accidentally used `stringify` in an invalid case.

You can use a linter to catch that error. [`typescript-eslint`](https://typescript-eslint.io/) supports type-aware linting rules. Even better, a linter will inform you that you've made a mistake. In an IDE, it will show you squiggly lines indicating a warning or an error.

Anyway, this is not a strong argument for having specialized return types. It's not worth the added complexity, especially when you have another tool that can solve this problem for you better.

> > If the return type is specialized to`string`, then the developer is lulled into a false sense of security. Sure, the result might actually be a `string`, but what if it isn't? As I described earlier, it is indeed possible for the result to be `undefined` because TypeScript is unsound.
> 
> That argument could be applied to literally any value:
> 
> ```ts
> function foo(x: string) {
>     // we say x is a string, but someone passed null to this function!
>     console.log(x.toLowerCase());
> }
> const a: string | null = null
> foo(a!)
> ```

Except, in the case of `JSON.stringify` we're only writing the type definition and not the implementation of the function. So, we have to be extra careful when writing the type definition because it won't be verified by the compiler. Thus, it's easy to have edge cases which are incorrect. Going back to my earlier example, stringifying any object with `toJSON` could potentially return `undefined`.

```typescript
const foo = JSON.stringify({ toJSON: () => {} });
const bar = JSON.stringify({ toJSON: () => () => {} });
const baz = JSON.stringify({ toJSON: () => Symbol() });
```

Writing the type definitions for handling all the edge cases will be very complex. So, even if you do get it correct, which is a very big if, you still need to think whether it's worth all the effort. IMHO, it's not worth the effort.

Just keep it simple and return `string | undefined`. You get added safety, and it's not much of an inconvenience for developers who are confident that the return type is only `string`. If you don't want to use the non-null assertion operator because it's unsafe, you could always just throw an error when you get `undefined`.

```typescript
const maybeString = JSON.stringify(someValue); // string | undefined

if (typeof maybeString === "undefined") {
  throw new TypeError("JSON.stringify failed to produce a string");
}

// The type of `maybeString` has been narrowed to `string`.
```

> > As you can see, it's easier to make the return type always be `string | undefined` than it is to create specialized type definitions for all the possible edge cases.
> 
> I don't think it's a far-fetched idea to specialize this function, even the `toJSON` case could be accounted for in the conditional return type.

Please, be my guest. I would love to see the type definitions required for handling all the edge cases.

Comment (aaditmshah) [reply src/tests/json-stringify.ts:6]: @sstur @KotlinIsland Created my own PR for adding type definitions for `JSON.stringify`, #124.

Would greatly appreciate any and all feedback.

**Comment by aaditmshah** [comment]
@KotlinIsland Added support for return type specialization in my PR, #124. It also correctly handles `toJSON`, constructor functions, and symbols.

@sstur I also copied your test cases into my PR and all of them pass. In addition, I added several more test cases for the `replacer` function. If you're happy with the changes that I made, we could close this PR in favor of mine.

File: src/tests/json-stringify.ts (hunk 5 above)
Comment (DetachHead) [reply src/tests/json-stringify.ts:6]: > 1. The type `Record<string, any>` uses `any`, which disables type checking. Change the type to `Record<string, unknown>` instead and you'll see the type error.

the example he was responding to used `Record<string, any>` 

> That being said, I can see how `Object.assign(() => {}, someRecord)` would have the type `(() => void) & Record<string, unknown>` which is indeed assignable to `Record<string, unknown>`. Thus, `Record<string, unknown>` could indeed be a function at runtime. All the more reason to not specialize the return type of `JSON.stringify`.

just because it shouldn't be specialized in that case doesn't mean it shouldn't be specialized in other cases. for example primitives such as `string` and `number` can be specialized to always return `string` because we know there's no way they could be functions

> You can use a linter to catch that error. [`typescript-eslint`](https://typescript-eslint.io/) supports type-aware linting rules. Even better, a linter will inform you that you've made a mistake. In an IDE, it will show you squiggly lines indicating a warning or an error.

i don't get why you'd use a linter for an error that the compiler can detect. imo we should aim to catch as many errors at compiletime as possible, to reduce the need for additional tools like eslint

> in the case of `JSON.stringify` we're only writing the type definition and not the implementation of the function. So, we have to be extra careful when writing the type definition because it won't be verified by the compiler.

i agree, but that doesn't mean we shouldn't do it at all. as long as we have tests and verify that each edge case is covered, i see no reason not to do it

> Writing the type definitions for handling all the edge cases will be very complex. So, even if you do get it correct, which is a very big if, you still need to think whether it's worth all the effort. IMHO, it's not worth the effort.

> Please, be my guest. I would love to see the type definitions required for handling all the edge cases.

i don't think it's that complicated at all. here's my attempt:
```ts
type JSONablePrimitive = number | string | boolean;

type JSONStringifyResult<T> =
  | string
  | (
      T extends JSONablePrimitive
        ? never
        : undefined extends T
          ? undefined
          : T extends { toJSON: () => JSONablePrimitive }
            ? never
            : T extends object
              ? undefined
              : never
    );
```
sure it looks gross, but if you're used to conditional types it's just four if statements and a union (and i'm sure it could be rewritten in a way that looks nicer. in fact it looks like you already did in #124). perhaps i missed something but i can't imagine it getting much more complicated than this.

Comment (aaditmshah) [reply src/tests/json-stringify.ts:6]: @DetachHead I agree. I changed my mind and solved it too, [2 days ago](https://github.com/total-typescript/ts-reset/pull/124/files#diff-3dfa61d45907024ba4aacbc6e522106744510ce75f2a90a904a9bcbab9968108R10-R14). Created a separate PR for it. See, #124. My PR improves both `JSON.parse` and `JSON.stringify`. It even provides strong types for the `reviver` in `JSON.parse` and the `replacer` function in `JSON.stringify`. Furthermore, it correctly handles values with the `toJSON` method. Finally, it passes all the tests in this PR. I would highly recommend that you check it out. Would love any and all feedback.

**Comment by denis-sokolov** [comment]
Allow me to add a vote of support for this. I have recently experienced a bug in our codebase because of the incorrectly lax type of JSON.stringify.

For what it’s worth, I recommend against always returning `string | undefined`. The amount of non-null assertion operators it has added in our codebase when we tried it was too large, and led to basically completely ignoring this strict type. It’s okay to have a few false negatives, but not all of them.

**Comment by aaditmshah** [comment]
@denis-sokolov Shameless plug. Consider using [`better-typescript-lib`](https://github.com/uhyo/better-typescript-lib) instead of `ts-reset`. [Version 2.3.0](https://github.com/uhyo/better-typescript-lib/blob/master/CHANGELOG.md#v230) of `better-typescript-lib` added strong type definitions for `JSON.parse`, `JSON.stringify`, and `fetch().json()`. It also added better type definitions for promises. I can vouch for these type definitions because I implemented and tested them myself.

I originally implemented these type definitions in `ts-reset`.

1. #121
2. #123
3. #124

However, my PRs were [summarily dismissed](https://github.com/total-typescript/ts-reset/pull/121#issuecomment-1480962069) by @mattpocock. No offense to him. He's a brilliant guy, and he's doing god's work by spreading the gospel of sound typing in TypeScript. But, he's not open to new ideas, which is why I'm doubtful of his open-source leadership skills.

On the other hand, the author of `better-typescript-lib`, @uhyo, is a genuinely nice guy who listens to others people's ideas. And, the second biggest contributor to `better-typescript-lib`, @graphemecluster, is a veritable type wizard, just like Matt. I've also seen him answer questions on https://es.discourse.group/. He's a smart guy.

Anyway, I'm not recommending `better-typescript-lib` to you just because of it's contributors. I'm recommending it to you because I honestly believe that it's superior to `ts-reset` in every metric.

1. It doesn't require users to define a [`reset.d.ts`](https://www.totaltypescript.com/ts-reset#get-started) ambient type declaration file. If you're using `npm` or `yarn`, then all you need to do is [install](https://github.com/uhyo/better-typescript-lib#installation) `better-typescript-lib` as a dev dependency, and you're golden. If you're using `pnpm`, as I do, then you will also need to [configure](https://github.com/uhyo/better-typescript-lib#with-pnpm) `pnpm` to hoist `@typescript/*` packages in order to use `better-typescript-lib`.
2. `ts-reset` doesn't actually replace the built-in type definitions. It just adds additional type overloads to functions. Hence, it's still possible to shoot yourself in the foot with `ts-reset`. On the other hand, `better-typescript-lib` [actually replaces](https://github.com/uhyo/better-typescript-lib#how-it-works) all the built-in type definitions with entirely new type definitions. It's able to do this because of an obscure [new feature](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#supporting-lib-from-node_modules) introduced by TypeScript 4.5. This means that you will only see the new and improved type definitions in your IDE if you use `better-typescript-lib`. If you use `ts-reset`, you will still see the old type definitions.

In conclusion, `better-typescript-lib` is objectively more well-designed than `ts-reset`. In addition, it provides strong type definitions for `JSON.parse`, `JSON.stringify`, and `fetch().json()` _today_. And last but not least, it's maintained by an excellent open-source leader, uhyo, who listens to his users and responds to issues and pull requests in a timely manner. Hence, you can expect to receive good support going forward.

**Comment by sstur** [comment]
> No offense to him. He's a brilliant guy [...]

Hmm, I wonder if the very next thing will be an offense 🤔

> he's not open to new ideas, which is why I'm doubtful of his open-source leadership skills. [Some other guy] is a genuinely nice guy who listens

I guess I should have seen that coming; whenever someone opens with "No offense" it's likely it will be followed by something offensive 🤣

Anyway, better-typescript-lib looks interesting, will check it out.

**Comment by aaditmshah** [comment]
> > No offense to him. He's a brilliant guy [...]
> 
> Hmm, I wonder if the very next thing will be an offense 🤔
> 
> > he's not open to new ideas, which is why I'm doubtful of his open-source leadership skills. [Some other guy] is a genuinely nice guy who listens
> 
> I guess I should have seen that coming; whenever someone opens with "No offense" it's likely it will be followed by something offensive 🤣

Thank you for the writing tip. Fixed. If you don't want to offend people, then never write "no offense". I'll make that a habit. 🙂 

Removed "No offense to him."
